akka-sample-kafka-to-sharding-scala/processor/src/main/scala/sample/sharding/kafka/UserEvents.scala

```diff
@@ -46,32 +43,42 @@ object UserEvents {
             runningTotal.copy(
               totalPurchases = runningTotal.totalPurchases + 1,
               amountSpent = runningTotal.amountSpent + (quantity * price)))
-        case GetRunningTotal(_, replyTo) =>
+        case GetRunningTotal(id, replyTo) =>
+          ctx.log.info("user {} running total queried", id)
           replyTo ! runningTotal
           Behaviors.same
       }
     }
   }
 
-  /*
-   * The murmur2 message extractor matches kafka's default partitioning when messages
-   * have keys that are strings
+  /**
+   * Asynchronously get the Akka Cluster Sharding [[ShardingMessageExtractor]]. Given a topic we can automatically
+   * retrieve the number of partitions and use the same hashing algorithm used by the Apache Kafka
+   * [[org.apache.kafka.clients.producer.internals.DefaultPartitioner]] (murmur2) with Akka Cluster Sharding.
    */
-  class UserIdMessageExtractor(nrKafkaPartitions: Int)
-      extends Murmur2NoEnvelopeMessageExtractor[Message](nrKafkaPartitions) {
-    override def entityId(message: Message): String = message.userId
+  def messageExtractor(system: ActorSystem[_]): Future[KafkaClusterSharding.KafkaShardingNoEnvelopeExtractor[Message]] = {
+    val processorConfig = ProcessorConfig(system.settings.config.getConfig("kafka-to-sharding-processor"))
+    KafkaClusterSharding.messageExtractorNoEnvelope(
```

Comment:
We need to make it clear just to do this once. Or maybe this can be some kind of extension?

Reply:
We could make it an extension.  This is currently called once per processor node.  If it's an extension would that ensure it would only be initialized once per cluster?